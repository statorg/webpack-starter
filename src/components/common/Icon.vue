<script>
const icons = {}
const requireIcon = require.context('@/assets/svg', false, /\.svg$/)
requireIcon.keys().forEach(fileName => {
  const iconConfig = requireIcon(fileName)
  // console.log(iconConfig.default || iconConfig)
  const { id, viewBox, url } = iconConfig.default || iconConfig
  const iconName = fileName.replace(/^\.\/|\.\w+$/g, '')
  const [width, height] = viewBox.split(' ').slice(-2).map(Number)
  icons[iconName] = { id: id.replace('-usage', ''), width, height, url }
})

// console.log(icons)

export default {
  name: 'Icon',
  functional: true,
  props: {
    name: {
      type: String,
      default: ''
    },
    left: Boolean,
    right: Boolean,
    img: Boolean
  },
  render (h, { props, data }) {
    const { id, width, height } = icons[props.name]
    const componentData = {
      class: {
        'icon': true,
        [`icon_${props.name}`]: !!props.name,
        'icon_left': props.left,
        'icon_right': props.right
      },
      style: {
        height: '1em',
        width: (width / height) + 'em'
      }
    }

    return (
      <svg {...data} {...componentData}>
        <use xlinkHref={`#${id}`} />
      </svg>
    )
    // return props.img
    //   ? (
    //     <img {...data} {...componentData} src={url} />
    //   )
    //   : (
    //     <svg {...data} {...componentData}>
    //       <use xlinkHref={url.replace('-usage', '')} />
    //     </svg>
    //   )
  }
}
</script>

<style lang="scss">
.icon {
  display: inline-block;
  vertical-align: middle;
  height: 1em;

  @at-root svg#{&} {
    fill: currentColor;
  }

  &_left {
    margin-right: 0.4em;
  }
  &_right {
    margin-left: 0.4em;
  }
}

@include vue(v-icon) {
  display: inline-block;
  vertical-align: middle;
  background-color: currentColor;
  width: 1em;
  height: 1em;
  opacity: 0.5;

  &[left] {
    @extend .icon_left;
  }
  &[right] {
    @extend .icon_right;
  }
}
</style>
