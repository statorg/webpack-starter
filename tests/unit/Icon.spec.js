import { shallowMount } from '@vue/test-utils'
import Icon from '@/components/common/Icon'

describe('Icon.vue', () => {
  it('should have icon_left class when left prop is passed as true', () => {
    const wrapper = shallowMount(Icon, {
      propsData: {
        left: true
      }
    })
    expect(wrapper.classes()).toContain('icon_left')
  })

  it('should have icon_right class when right prop is passed as true', () => {
    const wrapper = shallowMount(Icon, {
      propsData: {
        right: true
      }
    })
    expect(wrapper.classes()).toContain('icon_right')
  })

  it('should render and match snapshot', () => {
    const wrapper = shallowMount(Icon)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
