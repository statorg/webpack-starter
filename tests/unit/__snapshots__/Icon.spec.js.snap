// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon.vue should render and match snapshot 1`] = `
<svg class="icon" style="height: 1em;">
  <use xlink:href="#undefined"></use>
</svg>
`;
